CodeMirror v5.36.0 version update

BurningTreeC opened the pull request:
This updates cm to latest 5.35.1
Improves performance noticeable

BurningTreeC commented:
@all_ @Jermolene @pmario 

I've updated all changed files to the latest v5.35.1

I've also (re)added the "e.preventDefault()" I had added some time ago to the older plugin version

Would you welcome it if I add some of the very useful addons like the search&replace (with regex)?
I could add some docs&how-tos for them

Also, should I add **Erlang mode**  :grin: ?

all the best, Simon

BurningTreeC commented:
These are some interesting addons:

Merge/Diff: https://codemirror.net/demo/merge.html
Bi-directional Text: https://codemirror.net/demo/bidi.html
Search/Replace: https://codemirror.net/demo/search.html
Autocomplete: https://codemirror.net/demo/complete.html

I had the last two already working some time ago, it's easy to set up.
The Search/Replace is very useful if performant enough
Autocomplete is nice to have

The Merge/Diff uses the google-diff-match-patch library

Bi-directional text allows editing from right to left

pmario commented:
I would be ok with search and complete in the default edition, but the other 2 should be plugins.

 - Search/Replace: https://codemirror.net/demo/search.html
 - Autocomplete: https://codemirror.net/demo/complete.html

BurningTreeC commented:
@pmario Ok then I'll add Search/Replace and Autocomplete

1) The other 2 as additional tiddlywiki plugins, is that what you intend? Or just a how-to-add-them-if-you-want in the readme?

2) I've seen that you @pmario have improved the tiddlywiki.js highlighting for codemirror. Is it intended to be for text/x-tiddlywiki only or can it be used for vnd.tiddlywiki alike?

pmario commented:
>I've seen that you @pmario have improved the tiddlywiki.js highlighting for codemirror. Is it intended to be for text/x-tiddlywiki only or can it be used for vnd.tiddlywiki alike?

This is for TWclassic only. ...

Jermolene commented:
Thanks @BurningTreeC is this ready to merge?

BurningTreeC commented:
@Jermolene - yes it's ready!

pmario commented:
I think, the best way to include more auto-completing, would be a good description how to implement them. .. or some more additional plugins, if possible. .. 

I could live with "js, html, xml, css" ... I personally would have no use for "sql" .. That's why I think additional plugins would be best. 

How is is ... code size wise? with up to 100k minified I wouldn't care too much :)

BurningTreeC commented:
@pmario the whole plugin is about 988k right now

atm there is only the basic anyword-autocompletion

there are other completion engines, too ... like the tern completor, adding dropdowns etc. ...
see the demos on https://codemirror.net ... hidden in the [addon section](https://codemirror.net/doc/manual.html#addons) there are more demos for some features, you can search for demo there

I don't want to make decisions about what more to include in the plugin
You can tell me what you want in

You're right about minify - ing the thing

BurningTreeC commented:
the vim keymap alone adds 75kb to this, minified
I would ditch the emacs and vim keymaps tbh because I'm having a hard time imagining that someone seriously uses them.

pmario commented:
>@pmario the whole plugin is about 988k right now

yea, that's a size I personally wouldn't use as the default setting. ... I don't need vim settings. ... 

I think it depends on the users - preferences. .. So it should be single plugins if possible.

BurningTreeC commented:
@pmario ok good! I'll remove vim and emacs - I have to take a look if sublime is needed, could be the default - if not I'll ditch that, too

Then they'll be single plugin-plugins with instructions about how to add them to the cm config

The search&replace functionality also as separate plugin?

the overall potential is to save ~120kb moving the search&replace and the keymaps to separate plugins

430kb then. I don't know what's the best minifier to shrink it more

BurningTreeC commented:
@pmario @Jermolene 
this is ready to merge

I'll move the keymaps and autocomplete to separate plugins

BurningTreeC commented:
Nah, I guess the ReadMe and others need to be updated, too

I'll do that when the plugin-plugins are done

merge postponed

pmario commented:
hmmm. .. It seems you are pushing CM for TW5 as I did for TWc :) .. nice work!

BurningTreeC commented:
@pmario thanks!
I think now it's done, more features can be added easily if desired at some point

would you test?

BurningTreeC commented:
@pmario @Jermolene 

codemirror is stripped down to 292kb
there are some files left that are required by more plugins, so I left them in
outsourcing them has the potential of cutting ~30 - 40kb
that's the max

What do you think?

BurningTreeC commented:
Now it's really done. 280kb
I've grepped for what's needed by more cm plugins and left that in, what's included in the main plugin now is a good basic package I think

I made readmes for all the addons - installation should be self-explanatory (I hope)

pmario commented:
I will give it a try soon

BurningTreeC commented:
thanks, Mario! for the closetags addon to work with vnd.tiddlywiki, there's probably a tw5-mode needed
it works with text/html, would be a nice plus if it worked with tw5

But nevermind, I'm happy with this

BurningTreeC commented:
down at ~190kb

pmario commented on plugins/tiddlywiki/codemirror/config.tid:
```diff
@@ -7,6 +7,7 @@ type: application/json
-      "inputStyle": "textarea"
+      "inputStyle": "textarea",
+      "lineNumbers": false
   }
```
I think this should be lineNumbers: true

Otherwise the basic editor seems to be the same as core without the plugin.

BurningTreeC replied:
it's enough to just omit that, linenumbers are default
I wanted to add that so that the default is really simple and it doesn't show a linenumber in $:/ControlPanel input fields

I don't care much about it, just thought it makes the basic configuration more basic

pmario commented:
I think, importing new plugins, should also set the `config tiddler accordingly`. ... I'll have a closer look. 

I know CM and the config tiddler, but I'm struggling to see changes after I installed "base" ... because the only thing that's active is "matchBrackets". ... but only if the config is done right. ...

pmario commented:
I did add the search-replace now. ... That one is easy to check. ... It does something. ... And I'm already more confident ;)

pmario commented:
`F3` and `Shift-F3` for find-next and find-previous seem to work too. tested on windows. .. So the docs imo needs this info

BurningTreeC commented:
The base plugin provides just some methods needed by the search&replace, some keymap plugins and some highlighting modes... it also adds matchingbrackets, maybe matchingbrackets should go elsewhere

pmario commented:
.... I'm starting to get excited again :))

pmario commented:
>The base plugin provides just some methods needed by the search&replace, some keymap plugins and some highlighting modes... it also adds matchingbrackets, maybe matchingbrackets should go elsewhere

How big is "base" code wise?

BurningTreeC commented:
>  importing new plugins, should also set the config tiddler accordingly

If you have a look on that, I'm really glad. Editing json tiddlers seems not to be possible in an easy way so I made that copy-paste instruction

pmario commented:
I think the demoedition  should contain it already.

BurningTreeC commented:
> F3 and Shift-F3 for find-next and find-previous seem to work too. tested on windows. .. So the docs imo needs this info

I'll add that

> How big is "base" code wise?

~70kb

> I think the demoedition should contain it already.

you mean the base addon?

pmario commented:
>If you have a look on that, I'm really glad. Editing json tiddlers seems not to be possible in an easy way so I made that copy-paste instruction

Yea, the compy to clipboard button is really nice and simple, but the JSON "missing comma problem" bites us. ... Users will forget the comma at the end in the prev. line and it will cause trouble.

pmario commented:
>you mean the base addon?

yea and the config tiddler should contain the stuff, that it works. So imo it is a good starting point for new users. ... docs + config tiddler already set up, so they can be sure that it works.

BurningTreeC commented:
> Yea, the compy to clipboard button is really nice and simple, but the JSON "missing comma problem" bites us. ... Users will forget the comma at the end in the prev. line and it will cause trouble.

The other idea would be making different config files for each possible combination of addons
That would be... who's good at math?

BurningTreeC commented:
Yes it should, I'm missing ideas... I've seen a json editor somewhere recently - we could get ideas from that

BurningTreeC commented:
it's the TW5-MultiUser https://github.com/OokTech/TW5-MultiUser

it has a json editor in the ControlPanel - I thought there was an online demo somewhere

BurningTreeC commented:
I guess that json editor only works on node ... 

How about adding two data tiddlers (not json) - one for the require, one for the configuration

Then with the click of a button we can add new requirements to the require tiddler and new configurations to the configuration tiddler

Another "save" button then overwrites the config tiddler with the entries of both tiddlers in the correct json syntax

BurningTreeC commented:
Like:

Each the codemirror plugin and each addon has two tiddlers "require" and "configuration"
They're all tagged with $:/tags/CMConfig or $:/tags/CMRequire

list ... all tag[cmconfig] indexes[] ... get all indexes
list ... all tag[cmrequire] indexes[] ... get all indexes

put the whole thing together and save it to $:/config/CodeMirror

pmario commented:
>Yes it should, I'm missing ideas... I've seen a json editor somewhere recently - we could get ideas from that

no much  simpler, ..  I have an idea

pmario commented:
I did just add `autocomplete` without `css` `html `  and the others. ... So I got a RSOD .. 

IMO `"extraKeys "` is not really clear, where to copy it atm ... I did it wrong, in the first run. 

So building up the config on import ... IMO will be a key feature. If WE humans do it, it will be wrong ;)

BurningTreeC commented:
Yes - I'm curious about your idea now!

for the autocomplete, there's a notification (it should be there) when dependencies are missing, on top of the readme, but not more ...

I'll happily ditch that for autoconfiguration

pmario commented:
>I'll happily ditch that for autoconfiguration

no no .. I'm just commenting my progress, so we don't forget it. .. After I did it right once, I know what to do. ... So now is the only chance for me, to feel like a newbie .. and I like it ;)

pmario commented:
>Yes - I'm curious about your idea now!

I think CM will get it's own module - startup registration mechanism, so instead of "consuming" the config tiddler, the CM startup should "create" a config-tiddler. ... And the whole "require-section" imo should be gone.

BurningTreeC commented:
Ok then - I lean back and watch! ;)

BurningTreeC commented:
Just tell me when I can do something

BurningTreeC commented:
@pmario @Jermolene 

I've added the possibility to create a top/bottom panel using addPanel() - accepts top/bottom and a domNode to display

I thought that the toolbar could be wrapped inside, because of:

cm doesn't create an iframe and (sometimes, not always) looses focus when clicking a toolbar button (try clicking some buttons repeatedly with changing intervals) - I thought maybe such a panel solves that

the panel would be sticky on top and could be open/closed

it seems easy to do

BurningTreeC commented:
For the toolbar buttons and the textarea sometimes loosing focus:

CM has some methods to handle when it looses focus:

"blur" (instance: CodeMirror, event: Event)
    Fires whenever the editor is unfocused.

BurningTreeC commented:
meta.js contains meta info for ALL modes in the CM distribution

I'd only leave those in that are provided as plugins / used within TW

BurningTreeC commented:
@pmario @Jermolene 

is it feasible to introduce an "AddOns" tab to the plugin library
making the codemirror addons plugin-type addon?

pmario commented:
I'll have a closer look at the stuff today

pmario commented:
It seems: https://github.com/Jermolene/TiddlyWiki5/pull/3169/commits/5dcce301c32c06dde5cb991e8bbc8b1c47049d80   does init a second CM instance, with engine.js. ... What does the panel do?

BurningTreeC commented:
>It seems: 5dcce30 does init a second CM instance, with engine.js. ... What does the panel do?

I've added that to be able to see if the toolbar could be integrated in such a panel ([demo](https://codemirror.net/demo/panel.html))... maybe solving the issue that clicking toolbar buttons sometimes causes the CodeMirror textarea to loose focus

The normal text editor does not loose focus because of the iframe it's embedded into

I wanted to remove it before this gets merged

BurningTreeC commented:
... if unused

BurningTreeC commented:
@pmario ... to clarify: the panel right now does nothing, sorry for the confusion

pmario commented:
> ... to clarify: the panel right now does nothing, sorry for the confusion

The -panel plugin inits a new CM editor instance. So it uses a lot of memory. .. I'm not sure, if it should be done that way. .... IMO the panel should use the existing CM instance.

BurningTreeC commented:
@pmario I think you understand the code better than I do - I just copied this from the CodeMirror repo, haven't looked at it ... I wanted to do something with it if I find the time...

Are you trying to add a panel? If you can point me to the line(s) to change so that it uses the existing CM instance, I'll change it instantly

pmario commented:
>maybe solving the issue that clicking toolbar buttons sometimes causes the CodeMirror textarea to loose focus

That's possible. .. CM doesn't use the IFrame trick, that TW default editor uses. ... BUT IMO this should be its own PR.

BurningTreeC commented:
@pmario 

> That's possible. .. CM doesn't use the IFrame trick, that TW default editor uses. ... BUT IMO this should be its own PR.

Ok, you mean adding the panel and wrapping the toolbar into it -> separate PR?

pmario commented:
>Ok, you mean adding the panel and wrapping the toolbar into it -> separate PR?

yes, I think so

BurningTreeC commented:
Ok

pmario commented:
I'm modifying your patch-2 atm. So please keep calm for a day ;)

BurningTreeC commented:
:grinning: 

sure!

BurningTreeC commented:
> does closetags also work with TW syntax, if tiddler-type is not set to text/html?

no it doesn't - it completes TW widget tags if the type is text/html, but not with vnd.tiddlywiki 
That's something I wished to figure out

pmario commented:
ok ... 

`require` config section is eliminated ... 

now I'm looking into auto-config sensible defaults if an cm-addOn is imported. 

I'll send you a PR, that should then be merged into your patch2 branch. ...

BurningTreeC commented:
Great, thanks!

BurningTreeC commented:
@pmario ... about the configuration section:

there's the "inputStyle": "textarea" setting which I think should be a default, because for mobile browsers CM switches it to "contenteditable" if this is not set explicitely

contenteditable causes issues like triggering the import process (dropzone widget) when pasting or completely loosing focus when clicking a toolbar button

pmario commented:
>no it doesn't - it completes TW widget tags if the type is text/html, but not with vnd.tiddlywiki
That's something I wished to figure out

We have the same problem, that I had with CM and TWclassic on tiddlyspace. ... Basically we need 2 different mime-type flags. ... One for syntax highlighting and a second one, for the "real" text type. The meta.js file contains info about mime-types and CM-modes

In TWc I did use edit-type as an additional field, that did allow us, to define codemirror mixed modes. ... 

With TW5 every additional field, that is not a system-field as eg: created, will be shown in the editor. .. And if users mess with it, it will be broken :/ So we basically need a new "type-dropdown" that lets us define an edit-mode. 

The 3rd mode that's probably needed is "syntax highlighter mode" eg: 

`text/css` ... can be already used with autocomplete and highlighting.... but ... tiddlywiki allows us to define CSS tiddlers, that include eg: transclusions. ... but they also should be highlighted using CSS.

Jermolene commented:
@BurningTreeC @pmario thanks for all your work on this, it sounds like you've made a lot of progress.

I'd like to add a section to the release note that explains the changes to end users, so at some point it would be great to have a bullet list of all the improvements. Thanks!

BurningTreeC commented:
I don't have the durchblick about what's needed to do to make vnd.tiddlywiki syntax highlighting happen
Note that I have stripped down meta.js ... removed everything I found not needed... I'll post in a next message what's still in

>With TW5 every additional field, that is not a system-field as eg: created, will be shown in the editor. .. And if users mess with it, it will be broken :/

Making the field hidden and giving it a very unlikely name?

>text/css ... can be already used with autocomplete and highlighting.... but ... tiddlywiki allows us to define CSS tiddlers, that include eg: transclusions. ... but they also should be highlighted using CSS.

I see, that means if the type is vnd.tiddlywiki and the tiddler is tagged $:/tags/Stylesheet it needs a mixed css - vnd.tiddlywiki handler for highlighting and completion

The first steps would then be:

- making a vnd.tiddlywiki mode
- making a vnd.tiddlywiki - css - mixed mode
- adding vnd.tiddlywiki and vnd.tiddlywiki-css to meta.js

Is that correct?

BurningTreeC commented:
@pmario ...

Note for the Tags-Closing addon:
It needs the xml highlighting, css highlighting and javascript highlighting installed

@Jermolene 

> I'd like to add a section to the release note that explains the changes to end users, so at some point it would be great to have a bullet list of all the improvements. Thanks!

Yea I'll make such a list when this is done

pmario commented:
>It needs the xml highlighting, css highlighting and javascript highlighting installed

imo only `xml` mode is needed to get html-like tags ... eg: `<$list>`  works with xml only. ... but `text/html` mime type, which is only partially true.

pmario commented:
>I'd like to add a section to the release note that explains the changes to end users, so at some point it would be great to have a bullet list of all the improvements. Thanks!

Yea, I think Simon has put a lot of work into it. ... I'll try to make the plugins "plug & play"-able, with sensible default config options. ... But the options should be end-user editable. ..

I'm not sure, how to implement this, without the need to modify engine.js every time a new plugin is added, because atm. option handling is done in engine.js.

BurningTreeC commented:
... not wrong but too many

pmario commented:
@BurningTreeC  .. are there any dependency checks atm?

@Jermolene  ... Does the core have a dependency function? other than startup?

pmario commented:
@BurningTreeC ... I do have a version, that let's you drag & drop plugins from a "kitchen-sink" to empty, without any manual configuration needed. ... BUT ... there is no dependency checks atm. 

which causes the "keymap" plugins to brick the program, if dependencies are missing :/

 .. So we will need to improve the docs here 

"auto-config" is a bit hacky at the moment. 
I want to create a ControlPanel settings tab, that let's users change settings. ... Then I'll send a PR to you :)

BurningTreeC commented:
@pmario  - No, I haven't integrated any dependency checks and there aren't any from CM-side

> drag & drop plugins from a "kitchen-sink" to empty, without any manual configuration needed. ... BUT ... there is no dependency checks atm.

This sounds good! 
 could we use wikitext list-filtering to check for existing tiddlers that are needed (plugin-dependencies) and add module-type fields and json-configuration entries only if the dependencies are installed?

Nah, forget it, I should better wait for your PR to see the problems before suggesting

I'll be happy to wait and try

Jermolene commented:
> @Jermolene ... Does the core have a dependency function? other than startup?

No. A future version of the plugin library could bring in dependent plugins based on the (misnamed) `dependents` field that we already define for plugins.

pmario commented:
>No. A future version of the plugin library could bring in dependent plugins based on the (misnamed) dependents field that we already define for plugins.

I knew, there is something, but I couldn't remember the name :)

pmario commented on plugins/tiddlywiki/codemirror/engine.js:
```diff
@@ -13,14 +13,33 @@ Text editor engine based on a CodeMirror instance
+		callback(title,$tw.modules.execute(title));
+	});
+};
 
```
```
$tw.modules.forEachModuleOfType = function(moduleType,callback) {
	var modules = $tw.modules.types[moduleType];
	$tw.utils.each(modules,function(element,title) {
		callback(title,$tw.modules.execute(title));
	});
};
```

uups, this whole block should have been removed :/ I just copied it, as an "commented" example. ... but it seems I did undo the comments ..

BurningTreeC replied:
It's removed!

BurningTreeC commented:
@pmario this is awesome, much easier than I expected +1
I just added the matchBrackets config, I had matchbrackets moved to closebrackets ...

Your system is very simple and straight-forward, I like it!

The vim keymap is default on the demo page?
I just tried the system with my own testing wiki... everything worked fine

I like the solution with the codemirror module-type and the type: bool

BurningTreeC commented:
@pmario 
I'd like to roll a dice if showing linenumbers should be default or not,
I tend to not and I'd like to win :grin:

BurningTreeC commented:
... the nice people over at CodeMirror have just updated to 5.36.0
just noting this - when all is done I'd like to upgrade

BurningTreeC commented:
@pmario 

Do you have a ControlPanel Tab in work?
I think this is good to give it a go, what do you think?

I'll happily update the readmes with my red pen and add infos about the config tiddlers

pmario commented:
>I'll happily update the readmes with my red pen and add infos about the config tiddlers

That would be nice

pmario commented:
>Do you have a ControlPanel Tab in work?

no ... I think the existing one works. at least for me. 

>I think this is good to give it a go, what do you think?

yes .. when we have the readme's ... _and_ the ControlPanel: settings page.

BurningTreeC commented:
@pmario 
You mean adding the configuration options to the bottom of the $:/ControlPanel Settings tab?
I can do that tomorrow if you're not about to make it

BurningTreeC commented:
@pmario with my last additions the settings can be made at the bottom of the settings tab

the keyMap setting is missing, I think you had an idea for that

pmario commented:
@pmario with my last additions the settings can be made at the bottom of the settings tab

I wanted to make the settings dynamic, when the plugin is installed, it should also add the config section to the ControlPanel. 

You may have seen: config-language.multids  in the codemirror base plugin dir. ... I also want to make the stuff be translatable.

BurningTreeC commented:
@pmario - oops, I'm already adding settings at the bottom of the settings tab
I could just change tags

pmario commented:
>@pmario - oops, I'm already adding settings at the bottom of the settings tab
I could just change tags

As I wrote, I want to make it dynamic in a new Tab. So we shouldn't have a problem with merging. But I'm not 100% sure how it should work. ... I'll also want to have it translatable.

pmario commented:
ahhh, .. I didn't see, that you used the stuff already :)

BurningTreeC commented:
> I'll also want to have it translatable.

Me too, I'm using the config-language.multitids ... should I change hint and info to uppercase like its done in the core?
Also, I think in the core it uses "description" , not info

2) if you want to make a separate codemirror tab within the settings tab (I think that's better) , I would change the tags for the settings to $:/tags/ControlPanel/Settings/CodeMirror already

BurningTreeC commented:
@pmario 

one thing I haven't tested - installing addons:

How does your dependency-system handle if I try to install an AddOn without havin CM installed
Or if I install an AddOn having CM installed but that AddOn depends on another AddOn?

BurningTreeC commented:
... does it automatically install what's needed?

Jermolene commented:
Hi @BurningTreeC you probably know this, but v5.36.0 of CodeMirror has just been released

BurningTreeC commented:
@Jermolene yes I've seen that, I want to update when we're done here

BurningTreeC commented:
@pmario @Jermolene 

I've updated to v5.36.0

I was thinking about the syntax highlighting and auto-tag-closing for vnd.tiddlywiki
somehow for the vnd.tiddlywiki mimetype cm just needs to think it's text/html - then both should work
But then {{}} and [[]] or << >> may not be handled correctly

I think at some point we should create a codemirror mode for tw5

pmario commented:
>How does your dependency-system handle if I try to install an AddOn without havin CM installed
Or if I install an AddOn having CM installed but that AddOn depends on another AddOn?

There's no dependency handling atm.

pmario commented:
>I think at some point we should create a codemirror mode for tw5

Yea, ... We could use TWc as a base. ... BUT ... TW classic cm mode was modified by the CM-core people. They stripped my original down, to make it compatible with some core-css changes. ... 

The problem ... Many elements use "comment" as a highlighting color. ... which is wrong. .. So we would need to fix TWc and create a TW5 mode. 

I think TW5 mode has to be implemented as a mixed mode. .. It needs TW-syntax, html-syntax, css ... So it wont be simple, if done right. ... So it should be a new PR not this one :)

BurningTreeC commented:
> There's no dependency handling atm.

ok

I've added some lines in engine.js that make tag-closing and syntax-highlighting work for tw5 and twc
... just figuring out how this can be done, but it works
couldn't we just extend html-mode to highlight {{}} <<>> and [[]] , too?

Then I've added some "hidden settings"

- cursor blink rate
- indent unit
- tab size

They're not handled right ... I've set the cursorblinkrate config tiddler to type: bool ... that works
the other two are integer values, "bool" doesn't seem to work for them. When you have time @pmario , would you take a look?

pmario commented on plugins/tiddlywiki/codemirror/engine.js:
```diff
@@ -111,7 +111,11 @@ function CodeMirrorEngine(options) {
+  } else {
+    config.mode = options.type;
+  }
 	config.value = options.value;
```
I'd prefer to change meta.js with the right settings. So imo no hack would be needed. .. We modified meta.js anyway. So it should probably get a new name, since it isn't original anymore.

BurningTreeC replied:
that's good. I don't know the right settings, I tried before

tiddlymeta.js a good name?
I'll grep what's using it and rename when I'm back - then I'll also revert the hack

pmario commented on plugins/tiddlywiki/codemirror/engine.js:
```diff
@@ -135,7 +139,12 @@ function CodeMirrorEngine(options) {
-	this.cm.setOption("mode",type);
+  var self = this;
+  if (type == "text/vnd.tiddlywiki" || type == "text/x-tiddlywiki") {
+    self.cm.setOption("mode","text/html");
```
compare with `===`

pmario commented:
What do you think about a C9 instance, where we could code at the same time?

pmario commented:
>They're not handled right ... I've set the cursorblinkrate config tiddler to type: bool ... that works
the other two are integer values, "bool" doesn't seem to work for them. When you have time @pmario , would you take a look?

:)) I knew it would bit me, if I didn't implement `int` out of the box

BurningTreeC commented:
> What do you think about a C9 instance, where we could code at the same time?

that would be awesome, but I'm leaving this moment. I'll have to see how C9 works, never done that. But I'd be in!

> :)) I knew it would bit me, if I didn't implement int out of the box

Haha :grin:

pmario commented:
>that would be awesome, but I'm leaving this moment. I'll have to see how C9 works, never done that. But I'd be in!

I did revive my old account, which was set up already. .. I'll have a look, how invitation works.

BurningTreeC commented:
> I did revive my old account, which was set up already. .. I'll have a look, how invitation works.

cool. I'll send you my username when my account is set up

About the PR for making vnd.tiddlywiki mode and tag-closing work
Would you initiate that? I think you've already done work in that direction and you know the files and functions to look for, I'll jump in then

BurningTreeC commented:
For dependency checking:

there's the 
\define install-plugin-button()

in $:/core/ui/ControlPanel/Modals/AddPlugins

if we simply place a filter there that lists if a dependency field is given and then
action-sendmessage tm-load-plugin-from-library

that'd be an easy install-dependency-plugins too

BurningTreeC commented:
@pmario 

Meta.js isn't used anymore since your PR - did you intend so?

BurningTreeC commented:
@pmario just forget what I just said, now I've seen that it gets required

BurningTreeC commented:
@pmario 

I've made the settings tab and corrected the editor font and theme config stuff

What's left is 

1) dependency handling if you want to implement that 
2) handling of integers for the CM settings
3) testing

BurningTreeC commented:
..and meta.js now is tw-meta.js

BurningTreeC commented:
@Jermolene 

List of improvements:

- Upgrade version to 5.36.0
- Modularization into
    - Base plugin
         - basic lightweigth configuration, ~230kb
    - "AddOn" plugins
        - Search and Replace
        - Autocompletion
        - Fullscreen editing
        - Auto-Close brackets
        - Auto-Close tags
        - Sublime Text keymap
        - Emacs keymap
        - Vim keymap
        - Highlighting modes
            - Javascript
            - Xml
            - Html (htmlmixed)
           - Html (htmlembedded)
           - Css
           - Markdown
            - X-Tiddlywiki
- Automatic configuration
- ControlPanel configuration tab
    - line numbers
    - theme
     - editor fontfamily (monospace or sans-serif)
     - cursor settings
     - line wrapping
     - autocompletion on/off (if installed)

@pmario - anything missing?

BurningTreeC commented:
I'm done @pmario @Jermolene

pmario commented on plugins/tiddlywiki/codemirror/engine.js:
```diff
@@ -69,6 +69,10 @@ function getCmConfig() {
 					break;
+					case "integer":
+					  value = parseInt(tiddler.fields.text.trim());
+					  config.element = value;
```
`element` is a variable not an object element. see code above. it's extracted from the tiddler title.

pmario commented on plugins/tiddlywiki/codemirror/engine.js:
```diff
@@ -69,6 +69,10 @@ function getCmConfig() {
 						config[element] = value;
 					break;
+					case "integer":
+					  value = parseInt(tiddler.fields.text.trim());
```
Important: There needs to be the base parameter `parseInt(var, 10)` otherwise we will have problems!

pmario commented:
I think, the settings should be part of the corresponding plugins. .. So Setting is only there if the plugins is installed. ... + some info, that there may be more plugins, and where to find them.

pmario commented:
I'm also not sure, if we should really include all keymaps into the demo. 
May be there is a "drag & drop install your plugins from here"-version and a second
cm-demo version. 

I think some users may download the whole wiki and have at least 2 or even 3 keymaps as unwanted, payload. 

what do you think?

BurningTreeC commented:
@pmario 

> I think, the settings should be part of the corresponding plugins. .. So Setting is only there if the plugins is installed. ... + some info, that there may be more plugins, and where to find them.

they are already part of the plugins, but most settings are part of the base plugin
atm it's just the autocomplete setting that isn't part of base
this setting doesn't show up if autocomplete isn't installed

> I'm also not sure, if we should really include all keymaps into the demo.

we shouldn't ... the sublime text is the most complete and the easiest of all, I'd leave that in the demo.
We should update the cm demo start-tiddler about keymaps&features

pmario commented:
ControlPanel: Appearence: Theme Tweaks already has a Font setting. ... IMO we should use these settings!

pmario commented:
>we shouldn't ... the sublime text is the most complete and the easiest of all, I'd leave that in the demo.
We should update the cm demo start-tiddler about keymaps&features

I would be OK with that decision. ... And I think @Jermolene since he uses Sublime Text editor :)

I personally need my own settings, I want to have something that's similar to brackets.io

BurningTreeC commented:
> ControlPanel: Appearence: Theme Tweaks already has a Font setting. ... IMO we should use these settings!

It's been a mess for me using them correctly within the codemirror.css / styles.tid stylesheets, so I changed it simply to monospace or sans-serif to let the browser/system handle the font-family

There's a font-family entry in codemirror.css and one in styles.tid
As I said, I've struggled with transcluding the font-family there from ControlPanel - Appearance

> I would be OK with that decision. ... And I think @Jermolene since he uses Sublime Text editor :)

Ok then I'll change that 

> I personally need my own settings, I want to have something that's similar to brackets.io

I am fine with the standard cm settings
If your planning to make your personal keymap, now we can add it as AddOn ... that's cool

pmario commented:
We forgot the licensing. ... I only want to have 1 license tiddler in codemirror base which only contains the following text for codemirror

```
// CodeMirror, copyright (c) by Marijn Haverbeke and others
// Distributed under an MIT license: http://codemirror.net/LICENSE
```

and

```
Copyright (c) 2004-2007, Jeremy Ruston
Copyright (c) 2007-2017, UnaMesa Association

// Distributed under an BSD license: https://tiddlywiki.com/#License
```

Every "minified" cm-addOn needs the same codemirror header

BurningTreeC commented:
I'll add them licenses

pmario commented:
IMO ControlPanel: Settings: Codemirror should also contain a top link to: 

$:/plugins/tiddlywiki/codemirror/usage

pmario commented:
IMO  $:/themes/tiddlywiki/vanilla/options/sidebarlayout   should be

fluid-fixed

same as tiddlywiki.com.

pmario commented on plugins/tiddlywiki/codemirror/license.tid:
```diff
@@ -0,0 +1,15 @@
+title: $:/plugins/tiddlywiki/codemirror/license
+
+CodeMirror, copyright (c) by Marijn Haverbeke and others
+
```
hm, without the new-lines :)

BurningTreeC commented:
@pmario 

- controlpanel tab has link to usage on top
- license.tid is updated
- license headers added to addon files
- codemirror demo sidebar is fluid-fixed
- codemirror demo hellothere updated

not done: making cm use editor-font settings from ControlPanel-Appearance
bc I'm not able

pmario commented on plugins/tiddlywiki/codemirror/license.tid:
```diff
@@ -1,15 +1,8 @@
 title: $:/plugins/tiddlywiki/codemirror/license
 
 ~CodeMirror, copyright (c) by Marijn Haverbeke and others
-
```
hmm, I didn't think about the rendering problem.  This is what I wanted

```
"""
~CodeMirror, copyright (c) by Marijn Haverbeke and others
Distributed under an MIT license: http://codemirror.net/LICENSE

Copyright (c) 2004-2007, Jeremy Ruston
Copyright (c) 2007-2018, UnaMesa Association
Distributed under an BSD license: https://tiddlywiki.com/#License
"""
```

also see copyright Jeremy is 2018 now :)

pmario commented:
>not done: making cm use editor-font settings from ControlPanel-Appearance
bc I'm not able

We just need to use the same theme variable: `$:/themes/tiddlywiki/vanilla/settings/codefontfamily`
in $:/plugins/tiddlywiki/codemirror/styles

We should even use the same dialog. So if someone changes CM settings also ControlPanel: Appearence is changed.

pmario commented:
CodeMirror needed a setting, since it wasn't there in TW, as the first demo was written. .. Now TW has it's own editor settings. ...

pmario commented:
I did modify this, after you may have seen it :/ Sorry!

```
SiteTitle imo should be: ~CodeMirror Demo
Site Subtitle: ~CodeMirror plugins for ~TiddlyWiki5
```

BurningTreeC commented:
>I did modify this, after you may have seen it :/ Sorry!

no problem, I'll change it later

then I'll try using the $:/themes/tiddlywiki/vanilla/settings/codefontfamily in styles.tid (I'll check how it's done in vanilla base) and transcluding the Appearance - tweaks - codefontfamily input field or copying it over

pmario commented:
I think, what we do at the moment is "polishing" .. I also have some ideas, about "HelloThere" .. .. But I think, this can wait until the whole thing is merged. 

@Jermolene there's your list with the changes: https://github.com/Jermolene/TiddlyWiki5/pull/3169#issuecomment-375203549

@BurningTreeC ... Once you changed the font-family setting, I consider the project ... ready for merging and testing by the community. ....

pmario commented:
@BurningTreeC Once the code is merged to master, ... I can send you an invite for cloud9, so we could improve the docs together, probably at the same time ;)

BurningTreeC commented:
@pmario 
> Once the code is merged to master, ... I can send you an invite for cloud9, so we could improve the docs together, probably at the same time ;)

:+1: 

> Once you changed the font-family setting, I consider the project ... ready for merging and testing by the community. ....

:champagne:

pmario commented:
@Jermolene ... I think it is merge-able. ...

BurningTreeC commented:
@pmario 
I've seen now that c9 is an amazon "free for 12 months" thing - I'm not that sympathetic with amazons limited-time free services and I'm inclined not to sign up

What I always wanted to try is figuring out if beakerbrowser or a multi-user solution works for something like this - working together on docs for tw or similar - but it may still be a bit early for that
(I could set up beakerbrowser latest in an instant, I guess working-together would need to figure out how)

Am I wrong that with beaker's fork functionality this could be done?

pmario commented:
>I've seen now that c9 is an amazon "free for 12 months" thing - I'm not that sympathetic with amazons limited-time free services and I'm inclined not to sign up

I did beta testing, when they where not amazon. ... So imo no limit here. ... but I'm not sure

pmario commented:
For me it looks like this: 

![c9 io-billing](https://user-images.githubusercontent.com/374655/37776389-741dd578-2de5-11e8-9085-68ebedc5fc91.gif)

Did you register with: https://c9.io/  and use github to sign in?

pmario commented:
also see: https://c9.io/pricing

BurningTreeC commented:
Hm. I can't find an option to sign in with github.
I'll give them a chance to convince me if I can find that option

BurningTreeC commented:
No @pmario - I'm not gonna sign up there. After connecting with github they still want me to sign up for a cloud9 aws account, and they want credentials I don't want to give them.

Alternatives or the github way

BurningTreeC commented:
- https://teletype.atom.io/

pmario commented:
I can use this one: https://c9.io/login

BurningTreeC commented:
#3184
